README.md
# pstools

Tools for Propensity Score Analyses

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Build Status](https://travis-ci.org/dewittpe/pstools.svg?branch=master)](https://travis-ci.org/dewittpe/pstools)
[![Coverage Status](https://img.shields.io/codecov/c/github/dewittpe/pstools/master.svg)](https://codecov.io/github/dewittpe/pstools?branch=master)

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/pstools)](https://cran.r-project.org/package=pstools)

[![Licence](https://img.shields.io/badge/licence-GPL--2-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-2.0.html)
[![minimal R version](https://img.shields.io/badge/R%3E%3D-3.0.2-6666ff.svg)](https://cran.r-project.org/)

## Examples

## Install

### From CRAN
This package is currently *not* on CRAN.  Development is moving forward.

### Developmental
Install the development version of `pstools` directly from github via the 
[`devtools`](https://github.com/hadley/devtools/) package:

    if (!("devtools" %in% rownames(installed.packages()))) { 
      warning("installing devtools from https://cran.rstudio.com")
      install.packages("devtools", repo = "https://cran.rstudio.com")
    }

    devtools::install_github("dewittpe/pstools", build_vignettes = TRUE)

*NOTE:* If you are working on a Windows machine you will need to download and
install [`Rtools`](https://cran.r-project.org/bin/windows/Rtools/) before
`devtools` will work for you.



NEWS.md
# pstools 0.1.1

## Important Changes to API
* `propensity` has been deprecated
* `dstats` has methods for `glm` and `formula`

## New Features
* `dstats()` and `standardized_differences()` added to replace 
  `propensity()` (#6)
* Five propensity score based weights (#1) (#5)
* Added a `NEWS.md` file to track changes to the package.

# pstools 0.1.0
Moved code from a single research project into a publicly developing R package.



tests/testthat.R
library(testthat)
library(pstools)

test_check("pstools")


tests/testthat/test-propensity_summary.R
test_that("model family is detected",
          {
            fit <- glm(mpg > 20 ~ wt + cyl, data = mtcars)
            expect_error(propensity(fit), "binomial") 
          })

test_that("exposure is 0/1",
          { 
            fit <- suppressWarnings(glm(y ~ x, data = list(x = 1:10, y = runif(10)), family = binomial()))
            expect_error(propensity(fit), "0/1 integer vector")
          })

test_that("means", 
          {
            fit <- glm(mpg > 20 ~ wt + cyl, data = mtcars, family = binomial())
            obj <- propensity(fit)

            expect_equal(obj$key, c("cyl", "wt"))
            expect_equivalent(unclass(obj[2, c(2, 6)]), 
                              aggregate(wt ~ I(mpg > 20), data = mtcars, FUN = mean)[, 2]
                              )
          })


R/pride.R
#' Pediatric Respiratory Infection, Deutschland (pride)
#'
#' Data dealing with lower respiratory tract infections (LRTI) in n=3.078
#' infants and children aged less than three years in the observational study
#' Pri.DE (Pediatric Respiratory Infection, Deutschland) in Germany.
#'
#' @source This data set was aquireed from the \code{nonrandom} R package.
#' https://cran.r-project.org/web/packages/nonrandom/
#'
#' @format a data.frame with n = 3078 observations on 15 variables.
#' \describe{
#'   \item{PCR_RSV}{current respiratory syncytial virus (RSV) infection (1) or not (0)}
#'   \item{SEX}{gender: male (1) and female (2)}
#'   \item{ETHNO}{ethnic group: German (1), European Union (2) and others (3)}
#'   \item{FRUEHG}{preterm delivery (1) or not (0)}
#'   \item{RSVINF}{former RSV infection (1) or not (0)}
#'   \item{HERZ}{congenital heart defect (1) or not (0)}
#'   \item{REGION}{German region: South (1), East (2), West (3) and North (4)}
#'   \item{AGE}{age in years}
#'   \item{VOLLSTIL}{current breast feeding or longer than two months (1) or not (0)}
#'   \item{EINZ}{siblings (1) or not (0)}
#'   \item{TOBACCO}{passive tobacco smoke exposure at home (1) or not (0)}
#'   \item{EXT}{external care (1) or not (0)}
#'   \item{ELTATOP}{parental atopy (1) or not (0)}
#'   \item{SEVERE}{severe LRTI (1) or not (0)}
#'   \item{KRANKSUM}{number of diagnosed LRTI}
#' }
#'
#' @examples
#' library(ggplot2)
#' data(pride, package = "pstools")
#' 
#' glmfit <- stats::glm(PCR_RSV ~ SEX + RSVINF + REGION + AGE + ELTATOP + EINZ + EXT, 
#'                      data = pride, family = stats::binomial())
#' plotdata <- cbind(pride, propensity = qwraps2::invlogit(fitted(glmfit)))
#' 
#' ggplot2::ggplot(plotdata) + 
#' ggplot2::aes_string(x = "propensity", fill = "factor(PCR_RSV)") + 
#' ggplot2::geom_density(alpha = 0.5)
"pride" 


R/dstats.R
#' Descriptive Statistics
#'
#' Given a logistic regression model, generate the propensity weights for the
#' subjects (rows) in the data set and report the descriptive statistics for
#' each predictor in the regression model.
#'
#' @details
#' The regression model is expected to estimate the probability of an exposure
#' (Z = 1) given a set of predictors, X, i.e., Pr[Z = 1 | X].
#'
#' NOTE: for binary predictors coded as 0/1, such as male, the default action of
#' \code{dstats} will return a mean (sd), that is, if the \code{formula} for the
#' regression model is of the form
#' \code{y ~ male}, \code{dstats} will summarize the variable \code{male} as if it
#' was continous predictor. To get the summary of the proportion of males, and
#' females will be reported too, use a regression formula of the form \code{y ~
#' factor(male)}.
#'
#' See \code{\link{psweights}} for details on the propensity score based weights
#' used by \code{dstats}.
#' 
#' @param x a regression object with class \code{glm}
#' @param ... ignored
#'
#' @return A \code{pstools_dstats} object which is a
#' \code{data.frame}, with descriptive statistics for each variable in and out
#' of the exposure group.
#'
#' @references
#' Li, Liang, and Tom Greene. "A weighting analogue to pair matching in
#' propensity score analysis." The international journal of biostatistics 9.2
#' (2013): 215-234.
#'
#' @seealso \code{\link[stats]{glm}} for fitting logistic regression models, or 
#' \code{\link[geepack]{geeglm}} for fitting GEEs.
#' \code{\link{psweights}} for the weights.
#'
#' @examples
#'
#' data(pride)
#' glmfit <- stats::glm(PCR_RSV ~ SEX + RSVINF + REGION + AGE + ELTATOP + EINZ + EXT, 
#'                      data = pride,
#'                      family = stats::binomial())
#' dstats(glmfit)
#'
#' @export
dstats <- function(x, ...) { 
  UseMethod("dstats")
}

#' @export
dstats.glm <- function(x, ...) {
  if (stats::family(x)$family != "binomial") {
    stop("expected binomial family regression model.")
  }

  # propensity scores
  ps <- qwraps2::invlogit(stats::fitted(x))

  dstats.formula(stats::formula(x), cbind(x[["data"]], ps = ps), ps) 
} 

#' @export
dstats.default <- function(x, ...) {
  message("Unknown class for dstats")
}

#' @export
#' @rdname dstats
#' @param formula a regression formula of the form \code{exposure ~ predictors1 + predictors2} 
#' @param data a \code{date.frame}
#' @param ps the bare column name in \code{data} indicating the propensity scores
dstats.formula <- function(formula, data, ps, ...) { 
  lhs_vars <- all.vars(lazyeval::f_lhs(formula))
  rhs_vars <- all.vars(lazyeval::f_rhs(formula))

  exposure_col <- lhs_vars
  ps_col       <- deparse(substitute(ps))

  mm <- lapply(c(lhs_vars, rhs_vars, ps_col),
               function(x) { stats::as.formula(paste("~", x, "+0")) }) 
  
  mm <- lapply(mm, stats::model.matrix, data = dplyr::select_(data, .dots = c(lhs_vars, rhs_vars, ps_col)))
  mm <- do.call(cbind, mm)

  out <- psweights_(as.data.frame(mm), exposure_col, ps_col)

  out <- tidyr::gather_(out, key_col = "key", value_col = "value",
                        gather_cols = setdiff(colnames(mm), c(exposure_col, ps_col))
                        )

  out <- dplyr::group_by_(out, .dots = list( ~ key, lazyeval::interp( ~ z, z = as.name(exposure_col))))

  out <- dplyr::summarize_(out, 
                           .dots = list("unadj_mean" = ~ mean(value), 
                                        "unadj_var"  = ~ var(value),
                                        "adj_mean_IPW" = lazyeval::interp(~ sum(w * value) / sum(w), w = as.name("psw_IPW")),
                                        "adj_var_IPW"  = lazyeval::interp(~ sum(w * (value - adj_mean_IPW)^2) / (sum(w) - 1), w = as.name("psw_IPW")), 
                                        "adj_mean_ACE_Exposed" = lazyeval::interp(~ sum(w * value) / sum(w), w = as.name("psw_ACE_Exposed")),
                                        "adj_var_ACE_Exposed"  = lazyeval::interp(~ sum(w * (value - adj_mean_ACE_Exposed)^2) / (sum(w) - 1), w = as.name("psw_ACE_Exposed")),
                                        "adj_mean_ACE_Unexposed" = lazyeval::interp(~ sum(w * value) / sum(w), w = as.name("psw_ACE_Unexposed")),
                                        "adj_var_ACE_Unexposed"  = lazyeval::interp(~ sum(w * (value - adj_mean_ACE_Unexposed)^2) / (sum(w) - 1), w = as.name("psw_ACE_Unexposed")),
                                        "adj_mean_ACE_MostInfo" = lazyeval::interp(~ sum(w * value) / sum(w), w = as.name("psw_ACE_MostInfo")),
                                        "adj_var_ACE_MostInfo"  = lazyeval::interp(~ sum(w * (value - adj_mean_ACE_MostInfo)^2) / (sum(w) - 1), w = as.name("psw_ACE_MostInfo")),
                                        "adj_mean_ACE_MWP" = lazyeval::interp(~ sum(w * value) / sum(w), w = as.name("psw_ACE_MWP")),
                                        "adj_var_ACE_MWP"  = lazyeval::interp(~ sum(w * (value - adj_mean_ACE_MWP)^2) / (sum(w) - 1), w = as.name("psw_ACE_MWP")) 
                                        )) 
  out <- dplyr::ungroup(out)

  attr(out, "continuous")   <- as.integer(unique(out$key) %in% names(data))
  attr(out, "exposure_col") <- exposure_col
  attr(out, "ps_col")       <- ps_col
  class(out) <- c("pstools_dstats", class(out))
  out 
}



R/psweights.R
#' Propensity Score Weights
#'
#' Five different matching weights based on an exposure and propensity score.
#'
#' @details
#' Let p denote the propensity score for a subject.  The five weights are as follows.
#' 
#' \tabular{llcc}{
#'                                                                          \tab                          \tab Exposed        \tab Non-Exposed        \cr
#' Average causal effect in study population                                \tab \code{psw_IPW}           \tab 1/p            \tab 1/(1-p)            \cr
#' Average causal effect in exposed                                         \tab \code{psw_ACE_Exposed}   \tab 1              \tab p/(1-p)            \cr
#' Average causal effect in unexposed                                       \tab \code{psw_ACE_Unexposed} \tab (1-p)/p        \tab 1                  \cr
#' Average causes effect in population for which sample is most informative \tab \code{psw_ACE_MostInfo}  \tab 1-p            \tab p                  \cr
#' Average causal effect in mathcin weight population                       \tab \code{psw_ACE_MWP}       \tab min(p,(1-p))/p \tab min(p,(1-p))/(1-p) \cr
#' }
#'
#' @param .data a \code{data.frame}
#' @param exposure the bare name for the column within \code{.data} indicating
#' the exposure or non-exposure.  Expect this to be an integer, or coercible to
#' an integer with values of zero or one.
#' @param ps the propensity scores.  Expected values between 0 and 1.
#'
#' @return a \code{data.frame} with the exposure and ps vectors returned along
#' with five different weights.  See Details for information on the five
#' weights.
#'
#' @examples
#' glmfit <- stats::glm(PCR_RSV ~ SEX + RSVINF + REGION + AGE + ELTATOP + EINZ + EXT, 
#'                      data = pride,
#'                      family = stats::binomial()) 
#'
#' ourpride <- dplyr::mutate(pride, pp = qwraps2::invlogit(fitted(glmfit))) 
#'
#' psw <- psweights(dplyr::tbl_df(ourpride), PCR_RSV, pp)
#' psw
#' 
#' # Mirrored Histograms
#' plot(psw)
#' 
#' @references
#' Li, Liang, and Tom Greene. "A weighting analogue to pair matching in
#' propensity score analysis." The international journal of biostatistics 9.2
#' (2013): 215-234.
#'
#' @export
psweights <- function(.data, exposure, ps) {
  psweights_(.data, deparse(substitute(exposure)), deparse(substitute(ps)))
}

#' @param exposure_col a character string
#' @param ps_col a character string
#' @export
#' @rdname psweights
psweights_ <- function(.data, exposure_col, ps_col) {
  UseMethod("psweights_")
}

#' @export
psweights_.grouped_df <- function(.data, exposure_col, ps_col) {
  warning("grouped_df will be ungrouped and a rowwise operation will take place.", call. = FALSE)
  psweights_(dplyr::ungroup(.data), deparse(substitute(exposure_col)), deparse(substitute(ps)))
}

#' @export
psweights_.rowwise_df <- function(.data, exposure_col, ps_col) {
  warning("rowwise_df will be ungrouped.", call. = FALSE)
  psweights_(dplyr::ungroup(.data), deparse(substitute(exposure_col)), deparse(substitute(ps)))
}


#' @export
psweights_.data.frame <- function(.data, exposure_col, ps_col) {
  
  if (!all(.data[[exposure_col]] %in% c(0, 1))) {
    stop("Exposure Vector needs to be 0/1", call. = FALSE)
  }

  if (any(.data[[ps_col]] < 0 | .data[[ps_col]] > 1)) {
    stop("Propensity score < 0 or > 1 detected.", call. = FALSE)
  }

  if (any(.data[[ps_col]] == 0 | .data[[ps_col]] == 1)) {
    warning("Propensity score of 0 or 1 detected.", call. = FALSE)
  }

  out <-dplyr::rowwise(.data)
  out <- dplyr::mutate_(out,
                        .dots = list(
                                     "psw_IPW"           = lazyeval::interp( ~ 1/(z*ps + (1-z)*(1-ps)),                  ps = as.name(ps_col), z = as.name(exposure_col)),
                                     "psw_ACE_Exposed"   = lazyeval::interp( ~ z + (1-z)*ps/(1-ps),                      ps = as.name(ps_col), z = as.name(exposure_col)),
                                     "psw_ACE_Unexposed" = lazyeval::interp( ~ z * (1-ps)/ps + (1-z),                    ps = as.name(ps_col), z = as.name(exposure_col)),
                                     "psw_ACE_MostInfo"  = lazyeval::interp( ~ z * (1-ps) + (1-z) * ps,                  ps = as.name(ps_col), z = as.name(exposure_col)),
                                     "psw_ACE_MWP"       = lazyeval::interp( ~ min(c(ps, 1-ps)) / (z*ps + (1-z)*(1-ps)), ps = as.name(ps_col), z = as.name(exposure_col))
                                     )
                        )
  out <- dplyr::ungroup(out)
  attr(out, "exposure_col") <- exposure_col
  attr(out, "ps_col") <- ps_col
  class(out) <- c("pstools_psweights", class(out))
  out 
}


#' @export
#' @param x a \code{pstools_psweights} object
#' @param ... ignored
#' @describeIn psweights Mirrored histograms
plot.pstools_psweights <- function(x, ...) {

  dat <- tidyr::gather_(x, 
                        key_col = "method",
                        value_col = "value",
                        gather_cols = c("psw_IPW",
                                        "psw_ACE_Exposed",
                                        "psw_ACE_Unexposed",
                                        "psw_ACE_MostInfo",
                                        "psw_ACE_MWP"),
                        factor_key = TRUE) 

  ggplot2::ggplot() +
  ggplot2::aes_string(x = attr(x, "ps_col")) +

  ggplot2::geom_histogram(data = dplyr::filter_(x, .dots = lazyeval::interp( ~ z == 1, z = as.name(attr(x, "exposure_col"))))) +
  ggplot2::geom_histogram(data = dplyr::filter_(x, .dots = lazyeval::interp( ~ z == 0, z = as.name(attr(x, "exposure_col")))),
                          mapping = ggplot2::aes_string(y = "-..count..")) +

  ggplot2::geom_histogram(data = dplyr::filter_(dat, .dots = lazyeval::interp( ~ z == 1, z = as.name(attr(x, "exposure_col")))),
                          mapping = ggplot2::aes_string(weight = "value", fill = "method"),
                          alpha = 0.5
                          ) +
  ggplot2::geom_histogram(data = dplyr::filter_(dat, .dots = lazyeval::interp( ~ z == 0, z = as.name(attr(x, "exposure_col")))),
                          mapping = ggplot2::aes_string(weight = "-value", fill = "method"),
                          alpha = 0.5
                          ) +
  ggplot2::facet_wrap( ~ method, nrow = 1)
                     

}


R/gee_stepper.R
#' Stepwise GEE Model Selection
#'
#' Starting with a model of y ~ 1, use a forward/backward step selection process
#' to find the preferable fit.
#'
#' @param fit a \code{geeglm} object.  the response, id, data, corstr, ... will
#' be reused in all the the fits called by \code{gee_stepper}
#' @param upper A formula with the rhs containing all the variables that *could*
#' be the the model.
#'
#' @return the "best" model based on QIC from the stepwise selection process.
#'
#' @examples
#' 
#' data(pride)
#' library(geepack)
#' geefit <- geepack::geeglm(PCR_RSV ~ SEX + RSVINF + REGION + AGE + ELTATOP + EINZ + EXT, 
#'                   id = REGION, data = pride, family = stats::binomial())
#' gee_stepper(geefit, formula(geefit))
#' 
#' @export
gee_stepper <- function(fit, upper) {
  UseMethod("gee_stepper")
}

#' @export
gee_stepper.geeglm <- function(fit, upper) {
  preds     <- all.vars(as.list(upper)[[3]])
  preds_out <- preds
  preds_in  <- character(0)

  fit0 <- stats::update(fit, formula = . ~ 1)

  while(TRUE) {
    fits <-
      c(list(fit0),
        lapply(preds_out, function(p) {stats::update(fit0, formula = stats::as.formula(paste(". ~ . +", p))) }),
        lapply(preds_in,  function(p) {stats::update(fit0, formula = stats::as.formula(paste(". ~ . -", p))) })
        )

    minqic <- which.min(sapply(fits, MESS::QIC)[1, ])
    fit0 <- fits[[minqic]]

    status <-
      dplyr::data_frame(preds = sapply(fits, function(f) {as.character(stats::formula(f))[3]}),
                        qic   = sapply(fits, MESS::QIC)[1, ],
                        pick  = seq_along(fits) == minqic)

    preds_in  <- all.vars(as.list(stats::formula(fit0))[[3]])
    preds_out <- dplyr::setdiff(preds, preds_in)

    print(status)

    if (minqic == 1) {
      break
    }
  }
  fit0
}



R/standardized_differences.R
#' Standardized Differences
#' 
#' Calculate the standardized differences (percentages).
#'
#' @param x a logistic regression object, or a \code{pstools_dstats} object.
#' @param ... ignored at this time.
#'
#' @return A list of \code{data.frame}s each element of the list is for a
#' different weighting method.
#' 
#' @seealso \code{\link{psweights}}, \code{\link{dstats}}
#'
#' @examples
#' data(pride)
#' glmfit <- stats::glm(PCR_RSV ~ SEX + RSVINF + REGION + AGE + ELTATOP + EINZ + EXT, 
#'                      data = pride,
#'                      family = stats::binomial())
#' 
#' std_diff <- standardized_differences(glmfit)
#'
#' std_diff
#' 
#' # the plot is a ggplot2 object
#' plot(std_diff)
#'
#' plot(std_diff, adj_methods = "adj_ACE_MWP") + 
#'   ggplot2::ggtitle("ACE MWP vs Unadjusted")
#' @export
standardized_differences <- function(x, ...) {
  UseMethod("standardized_differences")
}

#' @export
standardized_differences.glm <- function(x, ...) {
  standardized_differences(dstats(x))
}

#' @export
standardized_differences.pstools_dstats <- function(x, ...) {
  sets <-
    list(dplyr::select_(x, ~ key, lazyeval::interp( ~ z, z = as.name(attr(x, "exposure_col"))), ~ unadj_mean, ~ unadj_var),
         dplyr::select_(x, ~ key, lazyeval::interp( ~ z, z = as.name(attr(x, "exposure_col"))), ~ adj_mean_IPW, ~ adj_var_IPW),
         dplyr::select_(x, ~ key, lazyeval::interp( ~ z, z = as.name(attr(x, "exposure_col"))), ~ adj_mean_ACE_Exposed, ~ adj_var_ACE_Exposed),
         dplyr::select_(x, ~ key, lazyeval::interp( ~ z, z = as.name(attr(x, "exposure_col"))), ~ adj_mean_ACE_Unexposed, ~ adj_var_ACE_Unexposed),
         dplyr::select_(x, ~ key, lazyeval::interp( ~ z, z = as.name(attr(x, "exposure_col"))), ~ adj_mean_ACE_MostInfo, ~ adj_var_ACE_MostInfo),
         dplyr::select_(x, ~ key, lazyeval::interp( ~ z, z = as.name(attr(x, "exposure_col"))), ~ adj_mean_ACE_MWP, ~ adj_var_ACE_MWP)
         )
  
  std_diffs <-
    lapply(sets,
           function(x) { 
             dplyr::summarize_(dplyr::group_by_(x, .dots = ~ key),
                               .dots = list("std_diff" = lazyeval::interp( ~ 100 * diff(w1) / sqrt(sum(w)/2), 
                                                                          w1 = as.name(names(x)[3]), w = as.name(names(x)[4])) 
                                            )) 
           })

  sets <-
    lapply(sets,
           function(x) {
             rtn <-
               dplyr::left_join(dplyr::filter_(x, .dots = lazyeval::interp( ~ z == 0, z = as.name(attr(x, "exposure_col")))),
                                dplyr::filter_(x, .dots = lazyeval::interp( ~ z == 1, z = as.name(attr(x, "exposure_col")))),
                                by = "key",
                                suffix = c(".z0", ".z1"))
             names(rtn) <- gsub(".*mean.*(\\.z\\d)", "mean\\1", names(rtn))
             names(rtn) <- gsub(".*var.*(\\.z\\d)", "var\\1", names(rtn))
             dplyr::select_(rtn, .dots = list(lazyeval::interp(~ - z.z0, z.z0 = as.name(paste0(attr(x, "exposure_col"), ".z0"))),
                                              lazyeval::interp(~ - z.z1, z.z1 = as.name(paste0(attr(x, "exposure_col"), ".z1")))
                                              ))
           })

  out <-
    mapply(dplyr::left_join,
           x = sets, 
           y = std_diffs, 
           by = "key", 
           SIMPLIFY = FALSE)
  names(out) <- c("unadj", "adj_IPW", "adj_ACE_Exposed", 
                  "adj_ACE_Unexposed", "adj_ACE_MostInfo", "adj_ACE_MWP")

  attr(out, "continuous") <- attr(x, "continuous")
  class(out) <- c("pstools_std_diffs", class(out))
  out
}

frmt_pstools_std_diffs <- function(x, ...) {
  mean_sd <- function(x, y) {
    paste0(qwraps2::frmt(x), " (", qwraps2::frmt(y), ")")
  }

  percent <- function(x) {
    paste0(qwraps2::frmt(x * 100, 1), "%")
  }

  x <- lapply(x, dplyr::mutate_, .dots = list("cnt" = ~ attr(x, "continuous")))

  x <- 
    mapply(function(dat, cnt) {
             dat <- dplyr::rowwise(dat)
             dat <- dplyr::transmute_(dat,
                                      .dots = list( 
                                                   key = ~ key,
                                                   unexposed = ~ if (cnt) { mean_sd(mean.z0, sqrt(var.z0)) } else { percent(mean.z0) },
                                                   exposed = ~ if (cnt) { mean_sd(mean.z1, sqrt(var.z1)) } else { percent(mean.z1) },
                                                   std_diff = ~ paste0(qwraps2::frmt(std_diff, 2), "%")
                                                   )
                                      )
             dat <- dplyr::ungroup(dat)
                  },
           dat = x, 
           MoreArgs = list(attr(x, "continuous")), 
           SIMPLIFY = FALSE)
}

#' @export
print.pstools_std_diffs <- function(x, ...) {
  x <- frmt_pstools_std_diffs(x, ...) 
  print(x)
}

#' @param adj_methods a character vector, names standardized differences
#' adjustment methods to include in the plot.
#' @export
#' @describeIn standardized_differences plotting method for
#' \code{pstools_std_diffs} objects.
plot.pstools_std_diffs <- function(x, adj_methods = c("adj_IPW", "adj_ACE_Exposed", "adj_ACE_Unexposed", "adj_ACE_MostInfo", "adj_ACE_MWP"), ...) { 
  dat <- dplyr::bind_rows(x, .id = "adjustment")
  dat <- dplyr::filter_(dat, .dots = ~ adjustment %in% c("unadj", adj_methods))
  dat$key <- factor(dat$key)
  levels(dat$key) <- rev(levels(dat$key))

  ggplot2::ggplot(dat) +
  ggplot2::aes_string(x = "std_diff", y = "key", 
                      color = "adjustment", shape = "adjustment") +
  ggplot2::geom_point() +
  ggplot2::geom_vline(xintercept = 0)
}


R/propensity.R
#' Propensity Summary
#'
#' Given a logistic regression model, generate the propensity weights for the
#' subjects (rows) in the data set.
#'
#' @details
#' The regression model is expected to estimate the probability of an exposure
#' (Z = 1) given a set of predictors, X, i.e., Pr[Z = 1 | X].
#'
#' NOTE: for binary predictors coded as 0/1, such as male, the default action of
#' propensity_summary will return a mean (sd), that is, if
#' propensity_summary(geeglm(y ~ male)) will summarize the variable male as if it
#' was continous predictor. To get the summary for the proportion of males, and
#' females will be reported too, use propensity_summary(geeglm(y ~
#' factor(male))).
#'
#' \code{weight_method} controls the weight given to each row of the data set.
#' Let ps = Pr[Z = 1 | X].  The default setting is \code{weight_method = 1}.
#' 
#' \describe{
#'   \item{\code{weight_method = 1}}{\code{w =  min(c(ps, 1-ps)) / (z*ps + (1-z)*(1-ps))}}
#'   \item{\code{weight_method = 2}}{\code{w = 1/(z*ps + (1-z)*(1-ps))}}
#'   \item{\code{weight_method = 3}}{\code{w = z + (1-z)*ps/(1-ps))}}
#' }
#' 
#' @param fit a regression object such as 
#' @param weight_method one of three methods for defining the weight for each
#' row of the data set.  See Details.
#'
#' @return A \code{pstools_propensity} object which is a
#' \code{data.frame} with columns summarizing each varaible used as a predictor in the
#' propensity model. The function will determine if each variable is a
#' categorical or continous variable. if the variable is continuous the mean (sd)
#' is returned.  If the variable is categorical the the percentage of each level
#' is returned.  Standard differences are reported as well. Unadjusted and
#' Matched Weight (see Li and Greene (2012)) values are reported.
#'
#' @references
#' Li, Liang, and Tom Greene. "A weighting analogue to pair matching in
#' propensity score analysis." The international journal of biostatistics 9.2
#' (2013): 215-234.
#'
#' @seealso \code{\link[stats]{glm}} for fitting logistic regression models, or 
#' \code{\link[geepack]{geeglm}} for fitting GEEs.
#'
#' @examples
#' \dontrun{
#' data(pride)
#' glmfit <- stats::glm(PCR_RSV ~ SEX + RSVINF + REGION + AGE + ELTATOP + EINZ + EXT, 
#'                      data = pride,
#'                      family = stats::binomial())
#' propensity(glmfit)
#' summary(propensity(glmfit))
#' plot(propensity(glmfit))
#' }
#' @export
propensity <- function(fit, weight_method = 1) {
  UseMethod("propensity")
}

#' @export
propensity.glm <- function(fit, weight_method = 1) {
  .Deprecated("use dstats and/or standardized_differences")

  if (!(weight_method %in% 1:3)) { 
    stop("weight_method needs to be 1, 2, or 3.")
  }

  if (stats::family(fit)$family != "binomial") {
    stop("expected binomial family regression model.")
  }

  # Exposure 1, no-exposure 0
  z <- stats::model.frame(fit)[[1]]
  if (!all(z %in% c(0, 1))) { 
    stop("Expected Outcome vector is a 0/1 integer vector.")
  } 

  # propensity scores
  ps <- qwraps2::invlogit(stats::fitted(fit))

  # get all the variable from the model fit and build the needed model matrix,
  # including the dumby variables for factors and character variables.
  av <- all.vars(stats::formula(fit))

  mm <- lapply(attr(stats::terms(stats::formula(fit)), "term.labels"),
               function(x) { stats::as.formula(paste("~", x, "+0")) }) 
  mm <- lapply(mm, stats::model.matrix, data = dplyr::select_(fit$data, .dots = av))
  mm <- do.call(cbind, mm)

  # build a summary data.frame and add weights
  out <- dplyr::bind_cols(dplyr::data_frame(z, ps), dplyr::as_data_frame(mm))
  out <- dplyr::rowwise(out)

  if (weight_method == 1) { 
    out <- dplyr::mutate_(out, .dots = list("w" = ~ min(c(ps, 1-ps)) / (z*ps + (1-z)*(1-ps))))
  } else if (weight_method == 2) { 
    out <- dplyr::mutate_(out, .dots = list("w" = ~  1/(z*ps + (1-z)*(1-ps))))
  } else {
    out <- dplyr::mutate_(out, .dots = list("w" = ~ z + (1-z)*ps/(1-ps)))
  } 

  out <- dplyr::ungroup(out)

  # Calculate the mean and variance for (unadjusted) values and adjusted values
  # out <- tidyr::gather(out, key, value, -z, -w)
  out <- tidyr::gather_(out, key_col = "key", value_col = "value",
                        gather_cols = c(colnames(mm)))

  out <- dplyr::group_by_(out, .dots = list( ~ key, ~ z))
  out <- dplyr::summarize_(out, 
                           .dots = list(~ mean(value), 
                                        ~ var(value),
                                        `mean(adjvalue)` = ~ sum(w * value) / sum(w),
                                        `var(adjvalue)`  = ~ sum(w * (value - `mean(adjvalue)`)^2) / (sum(w) - 1)))
  out <- dplyr::ungroup(out)

  # Split the data.frame by exposure and column bind
  out <- dplyr::left_join(dplyr::filter_(out, .dots = list( ~ z == 0)),
                          dplyr::filter_(out, .dots = list( ~ z == 1)),
                          by = "key",
                          suffix = c(".z0", ".z1"))
  out <- dplyr::select_(out, .dots = list( ~ -z.z0, ~ -z.z1))

  # calculate the standardized difference (%)
  # add a column to not if the variable is continous or a factor
  out <- dplyr::mutate_(out, 
                        .dots = list(
                       `unadj std diff (%)` = ~ 100 * (`mean(value).z1` - `mean(value).z0`)       / sqrt((`var(value).z1` + `var(value).z0`)/2),
                       `adj std diff (%)`   = ~ 100 * (`mean(adjvalue).z1` - `mean(adjvalue).z0`) / sqrt((`var(adjvalue).z1` + `var(adjvalue).z0`)/2),
                       `continuousvar` =  ~ as.integer(key %in% names(stats::model.frame(fit)))))

  class(out) <- c("pstools_propensity", class(out))
  out
}

#' @export
summary.pstools_propensity <- function(object, ...) {
  dplyr::summarize_(dplyr::rowwise(object),
                    .dots = list(
                                 key = ~ key,
                                 `unadj z1` = ~ if (continuousvar) { paste0(qwraps2::frmt(mean(`mean(value).z1`)), " (", qwraps2::frmt(sqrt(`var(value).z1`)), ")") } else {qwraps2::frmt(mean(`mean(value).z1`)*100, 1)},
                                 `unadj z0` = ~ if (continuousvar) { paste0(qwraps2::frmt(mean(`mean(value).z0`)), " (", qwraps2::frmt(sqrt(`var(value).z0`)), ")") } else {qwraps2::frmt(mean(`mean(value).z0`)*100, 1)},
                                 `unadj std diff (%)` = ~ qwraps2::frmt(`unadj std diff (%)`, 1),
                                 `adj z1` = ~ if (continuousvar) { paste0(qwraps2::frmt(mean(`mean(adjvalue).z1`)), " (", qwraps2::frmt(sqrt(`var(adjvalue).z1`)), ")") } else {qwraps2::frmt(mean(`mean(adjvalue).z1`)*100, 1)},
                                 `adj z0` = ~ if (continuousvar) { paste0(qwraps2::frmt(mean(`mean(adjvalue).z0`)), " (", qwraps2::frmt(sqrt(`var(adjvalue).z0`)), ")") } else {qwraps2::frmt(mean(`mean(adjvalue).z0`)*100, 1)},
                                 `adj std diff (%)` = ~ qwraps2::frmt(`adj std diff (%)`, 1)
                                 )) 
}

#' @export
plot.pstools_propensity <- function(x, y, ...) { 
  plotting_data <-
    dplyr::select_(x, .dots = list("key", "`unadj std diff (%)`", "`adj std diff (%)`"))

  plotting_data <-
    tidyr::gather_(plotting_data,
                   key_col = 'variable',
                   value_col = 'stddiff (%)',
                   gather_cols = list("unadj std diff (%)", "adj std diff (%)"))

  plotting_data$variable <- factor(plotting_data$variable)

  plotting_data$key <- factor(plotting_data$key)
  plotting_data$key <- factor(plotting_data$key, levels = rev(levels(plotting_data$key)))

  ggplot2::ggplot(plotting_data) +
  ggplot2::theme_bw() +
  ggplot2::aes_string(x = "`stddiff (%)`", y = "key") +
  ggplot2::geom_point(mapping = ggplot2::aes_string(color = "variable", shape = "variable")) +
  ggplot2::geom_vline(xintercept = 0) +
  ggplot2::theme(axis.title.y = ggplot2::element_blank())
}

